### PR TITLE
WIP: Revamps pyprep. An effort by NeurodataDesign at Johns Hopkins U

### PR DESCRIPTION
**Pull request by adam2392**
Addresses: #5 

Fixes include:

1. full PREP implementation
2. SK-learn style `fit()` pattern
3. sub-pipelines as functions and classes, to allow customization
4. unit/integration tests of said files
5. `black` formatting
6. Revamped travis.CI 
7. Includes notebook comparisons against MATLAB EEGLab's PREP

TODO:
- [ ] update whats_new.rst
- [ ] some additional testing using data I have w/ scalp EEG
- [ ] clean up the raw data files to see if instead they can just be imported via mne-python (?)

@sappelhoff hope to get your initial feedback before tackling some of these "house-keeping" issues. Aamna and Liang are also done w/ the course at this point, so they may or may not help in this last endeavor. So I think between the two of us though, we can get this to however you want it to look :)

### CODE REVIEW

**Comment by adam2392** [comment]
> * The `prep_demo.ipynb` does not run "out of the box" because it has a wrong path for the example file
>   
>   * would be nice to link to this example from the website in the form of a sphinx gallery like on MNE-BIDS. https://github.com/mne-tools/mne-bids/blob/0aff75c0d02a1c84925a7a212659cff1c01d3f29/doc/conf.py#L119 ... in the future we could then even make this example interactive using binder (similar to how it's done on mne-bids)

Yes I agree, the datasets should be sort of "imported" via MNE API. Should be easily fixed here. 

Can you advise what you mean for linking it in the sphinx gallery? Is all that is necessary, an explicit reference in that `conf.py` file? If not, what else should we do?

> * Running the test suite, I get 39 warnings, most of which are deprecation warnings. It would be worth looking into resolving these warnings

Agreed. I think these are mainly from the montage API being overhauled in the MNE versions.

> * the `Makefile` contains some non-working commands ... e.g., the `build-doc` command. I suggest to remove all unnecessary commands, keep the file minimal, and fix what's broken

Can erase as necessary.

> * Travis does not seem to run on this PR - any idea why?

I believe this should be a setting on your end? I am not a maintainer on this repo, so I can't access the travis, but I think there's a setting that enables PRs/branches to have travis run? If not, I'm not sure...

**Comment by sappelhoff** [comment]
> I believe this should be a setting on your end? I am not a maintainer on this repo, so I can't access the travis, but I think there's a setting that enables PRs/branches to have travis run? If not, I'm not sure...

I just checked, and Travis is running the PR ... it's just not showing on GitHub somehow. Perhaps that'll be fixed automatically once one more commit is pushed.

> I am not a maintainer on this repo

I just made you a maintainer, thanks again for contributing together with Aamna and Liang.

> Can you advise what you mean for linking it in the sphinx gallery? Is all that is necessary, an explicit reference in that conf.py file? If not, what else should we do?

yes, the conf.py file needs to be edited to contain a sphinx-gallery import and the settings ... and more importantly, the ipynb example would have to be converted to a py file.

Some links to show you what I mean:
- importing sphinx-gallery in conf.py
    - https://github.com/mne-tools/mne-bids/blob/4234280958775c445861d2c93b2b41d5c040391a/doc/conf.py#L6
    - https://github.com/mne-tools/mne-bids/blob/4234280958775c445861d2c93b2b41d5c040391a/doc/conf.py#L44
- making sphinx-gallery settings in conf.py
    - https://github.com/mne-tools/mne-bids/blob/4234280958775c445861d2c93b2b41d5c040391a/doc/conf.py#L119-L135
    - not that the "binder" settings can be ignored for now
- converting the ipynb to a python/rst format
    - https://github.com/mne-tools/mne-bids/blob/master/examples/convert_eeg_to_bids.py
    - basically, the markdown cells of the ipynb become commented "blocks" in the py/rst file

**Comment by adam2392** [comment]
@sappelhoff besides codecov, things are working out.

I converted the jupyter notebook to .py files, and did some adaptation of the conf.py files per your suggestion. I'm not too experienced w/ making docs the way you all do at MNE, but if there are still things missing, let me know!

Happy holidays ;)

**Comment by adam2392** [comment]
> hi @adam2392 thanks for your patience :-)
> 
> If you could:
> 
> 1. make the CIs pass
> 2. update what's new (https://github.com/sappelhoff/pyprep/blob/master/docs/source/whats_new.rst)
> 
> then I can merge this PR and deal with the rest.

Fixed the bug that pings users for input in the pytest. Updated the testing to make use of some fixtures that get thrown back and forth (e.g. mne.io.Raw and mne.Montage).

**Comment by sappelhoff** [comment]
and it's in! Thanks a lot everybody! :-)

May it help future researchers!

**Comment by sappelhoff** [comment]
@adam2392 I now fully integrated the code and updated the docs (see bottom of examples page: https://pyprep.readthedocs.io/en/latest/examples.html)

I decided against using binder + sphinx gallery for now, and instead I just ran the notebook on my machine, uploaded the ipynb output and provided a link to a notebook viewer:

https://nbviewer.jupyter.org/github/sappelhoff/pyprep/blob/master/examples/prep_demo.ipynb

Interestingly, the results now don't match so well anymore. Check the "results" section of the notebook, comparing the pyprep output and the "Matlab output".

I did not change any crucial code. What I did was:

- fix some links
- account for the fact that prep_pipeline takes a montage, not a string
- introduce a more complete channel renaming function ... the map that you used before did not completely work --> I don't know how you didn't get errors before ... perhaps these errors were only recently introduced in MNE :man_shrugging: 

Anyhow, before we can make a release, we should check the full prep pipeline's output again.

Until that happens, I think it's okay to have this as an unreleased "work in progress" feature, which is also fine :-)
